Copy on write reshape

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88774

Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

This PR adds a copy kwarg to reshape, which triggers a copy-on-write reshape. The intention is to eventually make this default. It also adds a warning for improper use of mutation before/after reshape call.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @EikanWang @jgong5

ezyang committed:
[WIP] Copy on write reshape

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc EikanWang jgong5

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc EikanWang jgong5

[ghstack-poisoned]

ezyang committed:
Update on "[WIP] Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc EikanWang jgong5

[ghstack-poisoned]

thiagocrepaldi reviewed (approved):
torch/onnx/symbolic_*.py LGTM

ezyang committed:
Update on "Copy on write reshape"


Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

This PR adds a copy kwarg to reshape, which triggers a copy-on-write reshape. The intention is to eventually make this default. It also adds a warning for improper use of mutation before/after reshape call.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc EikanWang jgong5

[ghstack-poisoned]

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -1327,8 +1327,10 @@ def _index_fill_reshape_helper(g: jit_utils.GraphContext, self, dim, index):
+def _reshape_helper(g: jit_utils.GraphContext, input, shape, allowzero=0, copy=False):
     shape = _maybe_get_const(shape, "is")
+    copy = _maybe_get_const(copy, "b")
+    assert not copy, repr(copy)
```
ONNX doesn't have concept of mutation. Should be fine to just ignore `copy`.

ezyang replied:
fancy, that'll be great

BowenBao reviewed (commented):
My comment is only on ONNX part, inline suggestion and otherwise LGTM.

ezyang commented:
@dzhulgakov's idea is that we test if a storage is dirty on `data_ptr` access in storage; that will get us cheap reads, making the warnings more accurate.
